diff -ruN keepalived-2.0.11.orig/configure keepalived-2.0.11/configure
--- keepalived-2.0.11.orig/configure	2019-01-06 18:52:33.000000000 +0000
+++ keepalived-2.0.11/configure	2019-01-18 02:41:48.810860630 +0000
@@ -862,6 +862,7 @@
 enable_libipset
 enable_nftables
 enable_libnl
+enable_track_process
 enable_strict_config_checks
 enable_hardening
 enable_optimise
@@ -1576,6 +1577,7 @@
   --disable-libipset      compile without libipset
   --disable-nftables      build without nftables support
   --disable-libnl         compile without libnl
+  --disable-track-process build without track-process functionality
   --enable-strict-config-checks
                           build with strict configuration checking
   --disable-hardening     do not build with security hardening
@@ -3682,6 +3684,11 @@
   enableval=$enable_libnl;
 fi
 
+# Check whether --enable-track-process was given.
+if test "${enable_track_process+set}" = set; then :
+  enableval=$enable_track_process;
+fi
+
 # Check whether --enable-strict-config-checks was given.
 if test "${enable_strict_config_checks+set}" = set; then :
   enableval=$enable_strict_config_checks;
@@ -5609,6 +5616,9 @@
     if test .$enable_nftables != .no; then :
   as_fn_error $? "enable-nftables requires vrrp" "$LINENO" 5
 fi
+    if test .$enable_track_process = .yes; then :
+  as_fn_error $? "enable-track-process requires vrrp" "$LINENO" 5
+fi
 
 fi
 if test .$enable_iptables = .no; then :
@@ -10563,7 +10573,9 @@
     CONFIG_OPTIONS="$CONFIG_OPTIONS BFD"
   fi
 
-    for ac_header in linux/cn_proc.h
+    if test .$enable_track_process != .no; then :
+
+      for ac_header in linux/cn_proc.h
 do :
   ac_fn_c_check_header_mongrel "$LINENO" "linux/cn_proc.h" "ac_cv_header_linux_cn_proc_h" "$ac_includes_default"
 if test "x$ac_cv_header_linux_cn_proc_h" = xyes; then :
@@ -10571,8 +10583,8 @@
 #define HAVE_LINUX_CN_PROC_H 1
 _ACEOF
 
-       SYSTEM_OPTIONS="$SYSTEM_OPTIONS CN_PROC"
-       HAVE_CN_PROC=Yes
+           SYSTEM_OPTIONS="$SYSTEM_OPTIONS CN_PROC"
+           HAVE_CN_PROC=Yes
 
 $as_echo "#define _WITH_CN_PROC_  1 " >>confdefs.h
 
@@ -10581,6 +10593,10 @@
 
 done
 
+
+else
+  CONFIG_OPTIONS="$CONFIG_OPTIONS DISABLE_TRACK_PROCESS"
+fi
 fi
  if test $VRRP_SUPPORT = Yes; then
   WITH_VRRP_TRUE=
